opm/input/eclipse/EclipseState/Grid/Box.hpp

```diff
@@ -17,61 +17,75 @@
   along with OPM.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-
 #ifndef BOX_HPP_
 #define BOX_HPP_
 
+#include <opm/input/eclipse/EclipseState/Grid/GridDims.hpp>
+
+#include <array>
 #include <cstddef>
-#include <limits>
+#include <functional>
 #include <vector>
 
 namespace Opm {
     class DeckRecord;
-    class EclipseGrid;
+}
 
-    class Box {
+namespace Opm
+{
+    class Box
+    {
     public:
+        using IsActive = std::function<bool(const std::size_t globalIdx)>;
+        using ActiveIdx = std::function<std::size_t(const std::size_t globalIdx)>;
 
-
-        struct cell_index {
+        struct cell_index
+        {
             std::size_t global_index;
             std::size_t active_index;
             std::size_t data_index;
 
-
-            cell_index(std::size_t g,std::size_t a, std::size_t d) :
-                global_index(g),
-                active_index(a),
-                data_index(d)
+            cell_index(std::size_t g,std::size_t a, std::size_t d)
+                : global_index(g)
+                , active_index(a)
+                , data_index(d)
             {}
 
-
-            /*
-              This constructor should is used by the global_index_list() member
-              which will return a list of *all* the cells in the box. In this
-              case the active_index will be set to the global_index. This is a
-              hack to simplify the treatment of global fields in the FieldProps
-              implementation.
-            */
-            cell_index(std::size_t g, std::size_t d) :
-                global_index(g),
-                active_index(g),
-                data_index(d)
+            // This constructor should is used by the global_index_list() member
+            // which will return a list of *all* the cells in the box. In this
+            // case the active_index will be set to the global_index. This is a
+            // hack to simplify the treatment of global fields in the FieldProps
+            // implementation.
+            cell_index(std::size_t g, std::size_t d)
+                : global_index(g)
+                , active_index(g)
+                , data_index(d)
             {}
         };
 
-        explicit Box(const EclipseGrid& grid);
-        Box(const EclipseGrid& grid , int i1 , int i2 , int j1 , int j2 , int k1 , int k2);
+        explicit Box(const GridDims& gridDims,
```

Comment:
No need for explicit

Reply:
> No need for explicit

True, but I want it.

Reply:
You'll have to explain that...?

Other than nitpicks, I think the overall intention of this is good.

Reply:
>>> No need for explicit
>> 
>> True, but I want it.
>
> You'll have to explain that...?

I don't want implicit conversion from "bunch of stuff" to `Box`.  There are preconditions on the values of `i1` vs. `i2` &c so I want the source code to explicitly say `Box { ... }` whenever we're forming objects of this type.  I agree with ABSL's "Tip of the Week" [#142](https://abseil.io/tips/142) regarding explicit constructors in this regard.